lib/tests/streamlit/runtime/state/session_state_test.py

```diff
@@ -1102,3 +1144,117 @@ def test_key_id_mapping_update(self):
         assert key_id_mapper.get_key_from_id("wid2") == "key2"
         assert key_id_mapper.get_id_from_key("key") == "wid3"
         assert key_id_mapper.get_key_from_id("wid") == "key"
+
+
+# region Multiple callbacks
+
+
+def test_per_key_callbacks_only_fire_for_changed_keys() -> None:
+    """Test that per-key callbacks only fire for changed keys."""
+    calls: list[str] = []
+
+    def cb_a() -> None:
+        calls.append("a")
+
+    def cb_b() -> None:
+        calls.append("b")
+
+    ss = SessionState()
+    wid = "w-json"
+    meta = WidgetMetadata(
+        id=wid,
+        deserializer=lambda v: v,
+        serializer=lambda v: v,
+        value_type="json_value",
+        callbacks={"a": cb_a, "b": cb_b},
+    )
+
+    # Register metadata
+    ss._set_widget_metadata(meta)
+
+    # Old state: {value: {a:1, b:2}}
+    ss._old_state[wid] = {"value": {"a": 1, "b": 2}}
+    # New state: {value: {a:1, b:3}}
+    ss._new_widget_state.set_from_value(wid, {"value": {"a": 1, "b": 3}})
+
+    ss._call_callbacks()
+    assert calls == ["b"]
+
+
+def test_json_trigger_aggregator_routes_to_named_callback() -> None:
+    """Test that JSON trigger values are routed to the correct named callback.
+
+    This test verifies that for widgets using `json_trigger_value`, the
+    `event` field in the JSON payload is used to look up and execute the
+    corresponding callback from the `callbacks` dictionary.
+    """
+    called: list[str] = []
+
+    def cb_click() -> None:
+        called.append("click")
+
+    def cb_submit() -> None:
+        called.append("submit")
+
+    ss = SessionState()
+    wid = "w-trig"
+    meta = WidgetMetadata(
+        id=wid,
+        # Convert JSON string from proto into a dict
+        deserializer=lambda s: json.loads(s) if s else None,
+        serializer=lambda v: v,
+        value_type="json_trigger_value",
+        callbacks={"click": cb_click, "submit": cb_submit},
+    )
```

Comment:
The `json_trigger_value` defines the full widget state as a trigger value, right? Is there a way to enable a widget to have both trigger and selection events (e.g., which is a major need for our dataframe)?